### DISCUSSION TITLE
🐺 Cairo wishlist 🦀

### DISCUSSION THREAD

**Discussion by AbdelStark** [starkware-libs/cairo#1714] [General]
Hey Cairo builders 🐺, let's discuss about what you would like to see for Cairo.
It can be:
- new features in Cairo
- quality of life improvements

Ultimately we would need to create issues for feature requests, but for the moment we don't have a clear process on how to handle external demands. So let's discuss those here and we will try to review / filter the requests and see if we can integrate some of them in the roadmap.

We appreciate your feedback and we will try to do our best to fit your needs!
![Abdel__StarkWare_a_wolf_wearing_Santa_Clauss_hat_in_Santas_slei_311ca34c-dc44-4a47-8ba1-076cf2a4feab](https://user-images.githubusercontent.com/45264458/211351155-6b83b83b-d152-43cb-a6d5-c34d399b0b44.png)

**Comment by AbdelStark** [comment]
Add `is_some` and `is_none` to `Option` type.

## Description
The [is_some](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some) and [is_none](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_none) methods return [true](https://doc.rust-lang.org/std/primitive.bool.html) if the [Option](https://doc.rust-lang.org/std/option/enum.Option.html) is [Some](https://doc.rust-lang.org/std/option/enum.Option.html#variant.Some) or [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None), respectively.
See https://doc.rust-lang.org/std/option/#querying-the-variant for the description of those functions.

## Proposition

Proposed signature for `is_some()`:
```cairo
fn is_some() -> bool
```

Proposed signature for `is_none()`:
```cairo
fn is_none() -> bool
```

**Reply by orizi** [reply]
these are now supported - you currently need to add `use option::OptionTrait;` and `use option::OptionTraitImpl;`
similarly for `Result`.

**Comment by maciejka** [comment]
Less verbose/cheaper struct storage updates
## Description
To read single attribute full struct needs to be read into memory, to update single struct attribute full struct needs to be rewritten. It is both more verbose than necessary and probably (depending on implementation) inefficient. More in depth problem description is here: https://hackmd.io/@RoboTeddy/BJZFu56wF#Concisely-update-large-structs.

**Reply by tarrencev** [reply]
+1, compiler level struct packing would be great too

**Reply by shramee** [reply]
This would be so sweet!

**Comment by maciejka** [comment]
Automaticaly generate trivial view functions for storage variables
## Description
Trivial storage related view functions take a lot of space and make contracts diffucult to read and understand. It should be possible to mark storage variables as public (similar to Solidity `public` modifier). As a result a view function should be generated. 

## Proposition
Instead of:
```
struct Storage { 
  value: felt,
}

#[view]
fn value() -> felt {
  value::read()
}
```
use `public` 'modifier':
```
struct Storage { 
  #[public]
  value: felt,
}
```

**Comment by maciejka** [comment]
`bytes32` equivalent
## Description
When working on cross chain apps, Cairo short strings are very inconvenient. Felt based short strings and their solidity counterpart `bytes32` are of diffrent lenght(251 vs 256) and of different encoding (left vs right padding). There is no easy conversion between the two. 

## Proposition
There should be a `bytes32` equivalent type available in Cairo. The usual operations: comparison, bit operators, shift operators, index access should be available. 

See:  https://docs.soliditylang.org/en/v0.8.17/types.html#fixed-size-byte-arrays

**Comment by caseywescott** [comment]
I would love to have /* */ multi-line commenting as I often need to add large blocks of text to contextualize the code I write.

**Comment by maciejka** [comment]
Support structs as `l1_handler` parameters

## Description
Currently when, for example Uint256 is send from l1 to l2 it needs to be manually converted to the desired type:
```
@l1_handler
func handle_deposit{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
    from_address: felt,
    amount_low: felt,
    amount_high: felt,
    sender_address: felt,
) {
    let amount = Uint256(low=amount_low, high=amount_high);
```
When there are more than one struct the code gets noisy.
   
## Proposition
When structs are supported as `l1_handler` parameters, above would be more readble: 
```
@l1_handler
func handle_deposit{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
    from_address: felt,
    amount: Uint256,
    sender_address: felt,
) {
```

**Comment by 0xAurelou** [comment]
It would be nice if we can have a float type. I know it's not as easy as people think to create but it would help a lot programmers :rocket:

**Reply by swapnilraj** [reply]
Do you mean an IEEE-754 float type, or do you mean a fixed point decimal type?
If its the first then I think it should be a library rather than a language built-in.

**Comment by Bethendawel123** [comment]
It's functionality can be improve

**Comment by florian-bellotti** [comment]
It would be awesome to have a cairo plugin for Intellij.

**Reply by mkaput** [reply]
@software-mansion team plans to work on this 😃

**Comment by swapnilraj** [comment]
StarkNet Cairo should allow function overloading. To make this possible, we will need to modify the function selector calculation to include the input types and the function name.
This change will affect type-checking and remote contract calls from the language perspective. However, resolving these is easy because we can statically determine which instance of the function is being used.
Since we can statically resolve function overloading, it doesn't introduce any performance overhead.

**Comment by gaetbout** [comment]
Adding an is_empty() fn on Array Trait could be very handy 
(I know I could do arr.len() == 0_usize, but isn't arr.is_empty() way more readable?)

**Reply by orizi** [reply]
added it to the trait now - available at main or at next release :)

**Comment by maciejka** [comment]
add `address` type

## Description
Currently `felt` type is used to represent contract addreses, e.g.:
```
   fn transfer_from(sender: felt, recipient: felt, amount: u256)
```
## Proposition
There should be an `address` type to represent contract addresses:
```
   fn transfer_from(sender: address, recipient: address, amount: u256)
```

**Reply by orizi** [reply]
we already have such a type `ContractAddress` - will change ERC20 impl to use it.
would also consider adding an alias `type address = ContractAddress;`

**Reply by delaaxe** [reply]
will this also be reflected in the ABI?

**Reply by Darlington02** [reply]
Does that mean this type can only be used on contract addresses? I mean the `ContractAddress` type?

**Reply by orizi** [reply]
* if the address is the type of the inputs - it'd be reflected in the ABI, of course a different abi can decide different strategies.

* you'd be able to construct it from `felt`s i guess - but its meaning is a contract-address.

**Reply by maciejka** [reply]
> we already have such a type `ContractAddress` - will change ERC20 impl to use it. would also consider adding an alias `type address = ContractAddress;`

`address` would be more in line with `felt`, `u256`. On the other hand, having to many conventions (CamelCase, `short types`?)  is not sth we want.

**Reply by milancermak** [reply]
The alias `type address = ContractAddress` is very handy, I'm for including it in the lang.

**Comment by gaetbout** [comment]
### Optimize usage of storage

Let's assume I have a struct:
`struct { some_field1: u8, some_field2: u32, some_field3: u64 }`

It should underneath be all contained in one felt rather than 3 separate felts to use the less possible storage on L1.

**Reply by mkaput** [reply]
I think this should be opt-in behaviour, like Rust's `#[repr(...)]`, because as far as I remember, this involves doing extra operations while reading/writing compressed fields?

**Reply by gaetbout** [reply]
I think it does indeed (lot of masks and bitwise operations), so depending on how optimized the costs for those is, it may or may not be more gas efficient...

**Reply by gaetbout** [reply]
IMO if when creating a struct, you can somehow specify the masks that need to be used to only perform the bitwise related operation it can already help.
But then it creates some burdens on the developper as he can make a mistake and not define a correct mask.

**Reply by mkaput** [reply]
That brings unsafety issues. I think this should be entirely automated and hidden from the programmer, so that a) we will have space to investigate various optimization patterns throughout time because there won't be any internal stability guarantees, and b) users won't be able to mess up things.

I think following would do the job:

```rust
#[repr(packed)]
struct Foo { some_field1: u8, some_field2: u32, some_field3: u64 }
```

or perhaps introduce versioning, because people will store this on the chain so there must be a possibility for backward and perhaps forward compatibility:

```rust
#[repr(packed, version = 1)]
struct Foo { some_field1: u8, some_field2: u32, some_field3: u64 }
```

**Reply by tarrencev** [reply]
This is handled by the `StorageAccess` trait, which will allow packing to be finely controlled. For example, here is how we store a custom struct in Dojo:

It could be extended to handle packing / unpacking and we can introduce procmacros to generate the logic

```
impl StorageAccessPosition of starknet::StorageAccess::<Position> {
        fn read(
            address_domain: felt, base: starknet::StorageBaseAddress
        ) -> starknet::SyscallResult::<Position> {
            Result::Ok(
                Position {
                    x: starknet::storage_read_syscall(
                        address_domain, starknet::storage_address_from_base_and_offset(base, 0_u8)
                    )?,
                    y: starknet::storage_read_syscall(
                        address_domain, starknet::storage_address_from_base_and_offset(base, 1_u8)
                    )?,
                }
            )
        }
        fn write(
            address_domain: felt, base: starknet::StorageBaseAddress, value: Position
        ) -> starknet::SyscallResult::<()> {
            starknet::storage_write_syscall(
                address_domain, starknet::storage_address_from_base_and_offset(base, 0_u8), value.x
            )?;
            starknet::storage_write_syscall(
                address_domain, starknet::storage_address_from_base_and_offset(base, 1_u8), value.y
            )
        }
    }
```

**Comment by tarrencev** [comment]
What do you think about providing access the contract context inside of starknets `StorageAccess` trait? I'm not sure exactly how that would look. Perhaps it is easier once we have the new contract syntax.

It would allow us to add access control directly at the storage level, which seems much more robust and less error prone.

One approach could be a `Ownable<T>` struct that wraps values but that would duplicate the owner state for each variable.

**Reply by spapinistarkware** [reply]
I think Ownable is a wonderful idea, and with some clever type engineering we can avoid the extra state.
Let me try,:
```rust
struct AdminToken{}
fn getAdminToken()->Option<AdminToken>{...}

#[storage]
struct MyContract{
  sensitive: LockedWith<felt, AdminToken>
}

fn get_sensitive(self: MyContract) -> Option<felt> {
  Some(self.sensitive.read().unlock(getAdminToken()?))
}
```

**Comment by tarrencev** [comment]
Support for `get_blockhash`

**Reply by clexmond** [reply]
This and / or a `get_random` potentially associated with the need for leader election randomness would be wonderful and a huge boost to games.

**Comment by gaetbout** [comment]
Possibility to import following Rust format:
`use some_crate::{something, something_else};`

**Reply by gaetbout** [reply]
There it is: https://github.com/starkware-libs/cairo/pull/2944 🥳

**Comment by EvolveArt** [comment]
Hey! 

Would love to have more features on `ArrayTrait` natively such as `split` or a `PartialEq` implementation too!
Right now if we want to create an array with elements in it we have to append them sequentially, we could have a `vec!` equivalent like`arr!` that makes it less verbose to write.

**Comment by delaaxe** [comment]
Array slices

**Comment by gaetbout** [comment]
A way to debug (or at least print it in the console) anything:
```
trait PrintTrait<T> {
    fn print(ref self: T);
}
```

Which could allow to do:
`get_caller_address().print();`

Could also automatically be generated on a struct with a macro just like:
`#[derive(Debug)]`

**Comment by enitrat** [comment]
Hey, here are a few suggestions after using Cairo 1 for a few weeks:
- Fixed-size arrays known at compile time would be awesome. More generally, macro_rules for metaprogramming that would reduce the amount of code to write and maintain going from
```
let arr = ArrayTrait::new();
arr.append(1);
arr.append(2);
```
to
```
let arr = array![1,2]
```

- A `clone` array method. If we want to avoid implicit copies for arrays, we should have a method to clone arrays in the core library. Eventually, we could also have other array common operations such as `slice`, `find`, etc directly in the core library instead of relying on third-party libraries.

- An updated syntax for structs. The main motivation here is that updating structs used passed as `ref` arguments in functions and having to redefine _every_  field is extremely verbose, even more so when it has to be done multiple times per function.
```
let struct2 = MyStruct {
    field1,
    field2,     
    ..struct1
};
```

- Default traits implementations. Some types share the same functionalities, and this would avoid a lot of redundancy. I'm also thinking that it could be used for smart contracts as well. All ERC20-enhanced contracts would have the default implementation of ERC20, plus a set of custom functions like `pause` for ERC20Pausable.

- Enhanced error logs for the borrow checker. When ownership is moved, it is not always clear which instruction is responsible and it is sometimes hard to debug.

- Import everything from a module with `*`. E.g. : `use super::*`

**Reply by clexmond** [reply]
These are all excellent and would eliminate multiple utils we had to write for Cairo 0.

**Comment by maciejka** [comment]
Add signed integer type e.g.: `s256`

**Reply by clexmond** [reply]
Agree. We have `felt` right now as the only signed type, but it is inconsistently signed. I.e. you can't rely on comparison operators, there is no `signed_div_rem`. Even the implementation of `unsigned_div_rem` in https://github.com/keep-starknet-strange/quaireaux/blob/main/src/utils.cairo#L43 relies on converting felts into `u128`. In the short term it would be great to get back `sign`, `abs_value`, `signed_div_rem` (and `unsigned_div_rem`) in corelib.

**Comment by punk5736** [comment]
Would it be possible to have a public compiler engineering channel on the Discord? Even if the channel is read-only for most with contributors getting write access (maybe after merging their first PR), I think it would be useful to follow the behind the scenes planning, design considerations, and debugging for people who want to dive in.

**Comment by rodrodros** [comment]
### Custom Implicits
Let users define Custom Implicits at `Cairo 1` level the same way `Cairo 0` where it was possible pass variables as implicit arguments as well as being able to "modify" implicit parameters inside function definitions

**Reply by orizi** [reply]
can you give a specific missing use case for these? aren't `trait` based `self` methods enough?

**Reply by spapinistarkware** [reply]
You can use ref arguments to modify them. Is this enough?

**Reply by rodrodros** [reply]
Yes, with `ref` arguments we can achieve the same behaviour, although is not our preferred solution, mainly because we defined a custom implicit and used it like a global var (which lived only during the function execution). The implicit param was used almost everywhere allowing us not to be constantly passing and returning it by hand (although with `ref`, returning the implicit param is no longer needed).

It is definitely a nice to have since a big chunk of our code depends on this custom implicit, but indeed, it is not a necessity.

Regarding the `trait` based `self` methods, I don't see right now how to apply them to my use case.

Edit: Yeah, trait based self methods are a viable solution

**Comment by maciejka** [comment]
Add a separate type for short strings for extra typesafety when that matters.

**Comment by 0x6e69636f736861726b** [comment]
Support for bytes and bytes32 in solidity

**Comment by maciejka** [comment]
Add `EthAddress` type and use it as a type of `from_address` of `l1_handler`.
For example, instead of:
```
#[l1_handler]
fn handle_deposit(from_address: felt252, account: ContractAddress, amount: u256) {
```
let's have:
```
#[l1_handler]
fn handle_deposit(from_address: EthAddress, account: ContractAddress, amount: u256) {
```
`EthAddress` will be usefull in other cases when communicating with L1.

**Reply by milancermak** [reply]
FWIW, there's an implementation of it in the [test data bridge](https://github.com/starkware-libs/cairo/blob/main/crates/cairo-lang-starknet/test_data/token_bridge.cairo#L14).

**Reply by maciejka** [reply]
I know, but it is a local declaration while it should be part of the Starknet extension and used in `l1_handler` signature. `felt252` type should not be used unless really necessary. Better name for it would be `felt252_and_I_promise_I_know_what_I_am_doing` ;-).

**Reply by milancermak** [reply]
100% agree :)

**Reply by milancermak** [reply]
Was done in #2913

**Reply by maciejka** [reply]
Now it should be used as a type of `from_address` of `#[l1_handler]`.

**Comment by wraitii** [comment]
I am a little confused about traits. They don't actually work like Rust traits, and some things are odd.

```
trait Something {
  fn DoSomething(self: SomeType) {}
}
```
This for example defines a trait for `SomeType`, but while the trait is _not_ generic it still needs a separate implementation. This is essentially just a way to add methods to a specific type. This does not match the rust name for `Trait`, nor really any general idea of what a trait is. It actually matches the rust struct `Impl` more.

One can also do this:
```
trait Something<T> {
  fn DoSomething(self:T) {}
}
```
This _does_ define a generic trait. However, implementing it must be done so:
```
impl thisdoesntmatter of Something<SomeType> {
  fn DoSomething(self:SomeType) {}
}
```
This is somewhat un-necessarily verbose because SomeType is repeated, and more importantly the `impl thisdoesntmatter of Something<SomeType>` doesn't actually seem to do anything, because Cairo associates the function via the type of `self`, not the name of the trait.  As a matter of fact, this works:
```
impl thisdoesntmatter of Something<u256> {
  fn DoSomething(self: felt252) {}
}

// Callsite: call a u256-trait on a felt252
Something::<u256>::do_stuff(34);
```
As a sidenote, you can define several methods with different types and only the last one is usable:
```
trait Something {
  // Ignored
  fn DoSomething(self: felt252) {}
  fn DoSomething(self: u256) {}
}
```

The name of the 'Impl' is generally useless, except when the Trait implementation needs to be imported. Which itself leads to some oddities:

1. An Impl doesn't seem to need importing if it's defined in the same module as the trait
2. It does need to be imported otherwise
3. But you don't actually need to import the implementation, but rather the module containing it (which probably explains point 1)

This makes the `impl X of Y` syntax completely redundant. You could just define free functions and, via UFCS, use them as methods on the relevant types. Adding in traits doesn't seem to do much:

```
fn DoSomething(self: felt252) {}
34.DoSomething()
```
and even generically:
```
fn DoSomething<T>(self: T) {}
```

As a final-final note, it seems possible to have conflicting implementations of a trait in different modules. Importing several of them will fail with an explicit error, but this seems somewhat dangerous when that depends on which `use` call the user makes.

---

I don't really have a very specific question except "what is the intent here?". It seems to me that traits should be made either more like Rust-traits, or less like them to fully embrace the UFCS-ness of the language

**Reply by spapinistarkware** [reply]
I tried to answer most of the things here, you are welcome to move the discussion there: https://github.com/starkware-libs/cairo/discussions/2744#discussion-5057835

**Comment by whatthedev-eth** [comment]
update: Made new issue #2786 for this:

Add aliases for imports, something like:

`use integer::u128_try_from_felt252 as u128_to_felt`

**Reply by gaetbout** [reply]
Done in https://github.com/starkware-libs/cairo/pull/2844 🎉

**Comment by notV4l** [comment]
bring back ** operator

**Comment by maciejka** [comment]
### Generate `DispatcherStorageAccess` implementation as part of #[abi] expansion  

What should be a preferred pattern to call another contract using address from storage variable? I mean I can store directly `ContractAddress`:
```
struct Storage { _token: ContractAddress }
```
but then I need to wrap it with `ERC20Dispatcher` on every call:
```
ERC20Dispatcher { contract_address: _token::read() }.transfer(...);
```
If instead of `ContractAddress` _token would store `ERC20Dispatcher`, then methods can called directly:
```
_token::read().transfer(...);
```
Which is less noisy and does not add any runtime overhead (as far as I understand). Right now it requires implementing `ERC20DispatcherStorageAccess` which if storing a Dispatcher instead of ContractAddress would become a preferred way, should be auto generated as part of `#[abi]` expansion.

**Reply by spapinistarkware** [reply]
Agreed

**Reply by maciejka** [reply]
https://github.com/starkware-libs/cairo/pull/2804

**Reply by milancermak** [reply]
What do you all think about generating `Into` for the `#[abi]` as well? 

```rust
fn foo(dest: ContractAddress) -> u256 {
    let token: ERC20Dispatcher = dest.into(); // instead of ERC20Dispatcher { contract_address: dest }
    token.balance_of()
}
```

**Reply by maciejka** [reply]
Agreed

**Reply by maciejka** [reply]
https://github.com/starkware-libs/cairo/pull/2948

**Comment by builtbyleo** [comment]
Is it possible to implement ecrecover in the core lib? (is this even possible on starknet?)

**Reply by spapinistarkware** [reply]
Yes. Open an issue please:)

**Comment by gaetbout** [comment]
A way to have better selector translation.
Instead of having:
```rust
const MY_FN_SELECTOR = 52247371237620501210265746714146606633180142029568650268503144952930738537; // my_fn
```

Having the possibility to do:
```rust
const MY_FN_SELECTOR = 'my_fn'.to_selector();
```
Maybe (idk) a type is even relevant?

**Reply by spapinistarkware** [reply]
I think this requires constexprs, which will take a while to do

**Comment by tserg** [comment]
Can we have a `continue` for loops?

[EDIT: also opened an [issue](https://github.com/starkware-libs/cairo/issues/2939)]

**Reply by spapinistarkware** [reply]
Done (by you) :)

**Comment by wraitii** [comment]
It seems to me a `Construct` trait like so would be useful:
```rust
trait Construct<T> {
    fn new() -> T;
}
```

Once we have that, we could additionally define:
```rust
fn new<T, impl TConstruct: Construct<T>>() -> T {
    Construct::new()
}
```
and write (assuming Array implements it of course):
```rust
let someArray: Array<u8> = new();
// or
let someArray = new::<Array<u8>>();
```

----

Alternative names:
```rust
trait Default<T> {
    fn default() -> T;
}
```
```rust
trait Make<T> {
    fn default() -> T;
}
```
```rust
trait New<T> {
    fn default() -> T;
}
```

**Reply by gaetbout** [reply]
https://doc.rust-lang.org/std/default/trait.Default.html 
👀

**Reply by spapinistarkware** [reply]
Yes, sounds like the default trait. Ill add an issue

**Comment by wraitii** [comment]
Unlike rust, but like C++ templates, Cairo 1 supports "Generic Specialization". The following code will work:
```rust
trait SomeGenericTrait<T> {
fn doSomething(self: T) {}
}
// Default implementation
impl GenericImpl<T> of SomeGenericTrait<T> {...}

// A 'partial' specialization available if T is Copy:
impl SpecializedImpl<T, impl tcopy: Copy<T>> of SomeGenericTrait<T> {...}

// An 'explicit/full' specialization for felt252 (which happens to be copy)
impl FeltImpl of SomeGenericTrait<felt252> {...}

fn main() {
	// This would be ambiguous: there are 3 possible implementations.
	// 214.doSomething();
	// But Cairo 1 allows selecting which you want by name:
	GenericImpl::doSomething(214)
	SpecializedImpl::doSomething(214)
	FeltImpl::doSomething(214)
}
```

For reference:
- Rust RFC on Impl specialization: https://github.com/rust-lang/rfcs/blob/master/text/1210-impl-specialization.md
- C++ "Partial ordering of Constraints", which is essentially what the rust RFC says could be implemented: https://en.cppreference.com/w/cpp/language/constraints#Partial_ordering_of_constraints


As the example shows however, Cairo 1 does not have any notion of ordering/specificity within those implementations.
Which means you will have to explicitly write the impl name every time there are two possible conflicting implementations in inference scope, in particular a general/default one and a fully-specialised one.
Furthermore, implementations are inferred within the scope of types & traits, which means it's pretty easy to 'pollute' the inference scope with implementations that are general.

This is, in my opinion, pretty annoying.

There are a few possible solutions:

- Implement specificity ordering like C++ or the Rust RFC proposes, and pick the most specific if there is one. This is the most magic, which has obvious pros and cons (namely explicitness)
- Allow declaring `fallback` implementations, that would never be selected if there are other possible implementations. This only really works to a point, but it would simplify the 'Generic implementation & fully-specialized case', which I'd expect to be the most common. This is similar to something described in the Rust RFC.
- Allow some syntax to hide implementations / prefer implementations. This would be useful if you're in code that needs a specific trait a lot and where the name of the impl is just noise. This could look like that:
```rust
// Proposed syntax:
use FeltImpl for SomeGenericTrait<felt252>;

fn main() {
	let afelt = 214;
	// No longer ambiguous: we have selected FeltImpl manually.
	214.doSomething();
}
```
There are IMO two questions. First, how to write this:
```rust
use SpecializedImpl for SomeGenericTrait<T> with Copy<T>;
```
Then 'should this be contagious across modules'
```rust
mod specialization {
    use SpecializedImpl for SomeGenericTrait<T> with Copy<T>;
}
use specialization;
fn main() {
	let afelt = 214;
	// No longer ambiguous: we have selected SpecializedImpl in the imported submodule. Or not.
	214.doSomething();
}
```

----

As an alternative, I propose the following 'Scoped default implementation' pattern:
```rust
mod some_trait {
    trait SomeTrait<T> {};
    mod generic_impl {
        impl genericImpl<T> of SomeTrait<T> {}
    }
}

// in user code:
use some_trait::SomeTrait;
impl feltImpl of SomeTrait<felt252>
fn main() {
    // Not ambiguous: only the felt impl is in inference scope
    1234.something();
}

// in another user code:
use some_trait::SomeTrait;
use some_trait::SomeTrait::generic_impl; // This puts the generic impl in inference scope
fn main() {
    // works via the generic impl
    1234.something();
}
```

**Reply by spapinistarkware** [reply]
I'm afraid of "magic", where the developer sees an impl, and it's the wrong one, and hr has no clue where to look for the right one, or even that it's wrong.

I want to add upon these suggestions: that we can only resolve ambiguity, when all the seen impls are in the same module, or local to our scope.

That way, if there is a local impl, it is apparent, and otherwise, the not chosen impls are always near the chosen impl.

**Reply by wraitii** [reply]
One drawback I see to not having 'automatic' specialization is recursive compile-time code.
Consider (assuming a future with const-felt-generics):
```rust
fn test<const a: felt252>() {
    test::<a-1>();
}
// In C++, you'd write this to break the recursion and make things compile.
// but in Cairo this will lead to ambiguity that cannot be resolved a priori.
fn test<0>() {}
```
The workaround would be to prevent the generic template from applying to 0 (horrible syntax example provided)
```rust
fn test<const a: felt252, constexpr { a > 0 }>() {
    test::<a->();
}
fn test<0>() {}
```

**Comment by rkdud007** [comment]
Current Dictionary cairo supports
```
mod dict;
use dict::Felt252Dict;
use dict::SquashedFelt252Dict;
use dict::felt252_dict_new;
use dict::felt252_dict_write;
use dict::felt252_dict_read;
use dict::felt252_dict_squash;
use dict::Felt252DictTrait;
```

------

So, only Dict type is `Felt252Dict<T>`. Because it its made by Felt252, we cannot put u256 in <T>. 
So like in memory, need u256 type. But we cannot do `Felt252Dict<u256>`. ( thats why in kakarot-ssj used StackU256helper ) 

I think would be useful if can use something like `dict::U256DictTrait;` from cairo directly :)

**Reply by makluganteng** [reply]
I agree it would be better for the felt252dict to support auto u256 when it receive a type of u256 it will be able to adjust. Mean while we need to create a helper function for it.

**Reply by spapinistarkware** [reply]
You mean you want a dict with the u256 as a key type, right?

For value type, you have Nullable

**Reply by rkdud007** [reply]
https://github.com/sayajin-labs/kakarot-ssj/blob/4afa1e90629811ed8ab58b5c20d46a7f26b25341/src/stack.cairo#L117

**Reply by rkdud007** [reply]
> You mean you want a dict with the u256 as a key type, right?
> 
> For value type, you have Nullable

actuayll want u256 as a value type. Want to add item that type is u256 inside the dictionary. And base on felt252dict, i cannot find Nullable as value :/

**Comment by spapinistarkware** [comment]


**Reply by makluganteng** [reply]
Owww , so as in not putting an initial type ?

**Reply by spapinistarkware** [reply]
What's an initial type? There is a key type and a value type

**Reply by makluganteng** [reply]
sorry i mean the value that i want is u256 
so like in this case i made it like Felt252Dict<u256> but this is not possible. so maybe i was thinking to able to support a dict that is u256 ?

**Reply by enitrat** [reply]
We could have `Felt252Dict<T, impl TSerde: Serde<T>>` where T would be serialized into a `Array<felt252>` and then we would be able to loop to insert all the values `v_i` in the dict. But for this we need info on the size of `T` at compile time

**Reply by spapinistarkware** [reply]
I think I already said it: Use Nullable<u256> as the value type.

**Comment by dedubluman** [comment]
Can you add the ERC721-A standard to the Cairo contracts? This standard is highly needed for us NFT collectors, especially since gas fees are high on Starknet. It would be a lifeline for the NFT community in this regard. I would appreciate it if you could convey my request to the authorized individuals responsible for these matters.

**Comment by hgedia** [comment]
It would be great if functions like mulmod and addmod from would have built-in lib implementations or u256 types.

**Comment by sgc-code** [comment]
Ability to read the current class hash from a smart contract. We can change if but we can't read the value now

**Reply by augustbleeds** [reply]
yes please 👍

**Comment by sgc-code** [comment]
default (fallback) method, like it was in Cairo 0. Allows to implement diamonds, and beacon proxies among other things.

**Comment by FrostStarBook** [comment]
Add String support, and some common APIs of String

## Description
Adding support for String and some commonly used String APIs would greatly improve development efficiency, especially when developing ERC721 contracts. Currently, we are rewriting the ERC721 contract written in Solidity using Cairo. However, since there is no API for string concatenation, we have to convert it into an array to achieve the desired functionality. Additionally, the felt252 length is too short, so we also have to use an array to implement our URI.


## Proposition
I propose adding support for the String data type and introducing APIs for String manipulation, such as concatenation and substring operations. This would greatly enhance the development experience and streamline the process of working with strings in ERC721 contracts.

**Reply by gaetbout** [reply]
This might already be ongoing:
https://github.com/orgs/starkware-libs/projects/1/views/1?pane=issue&itemId=28823165

**Comment by FrostStarBook** [comment]
Arrays can be defined in the constant pool and support initialization data.

## Description
We are using Cairo to rewrite a Solidity ERC721 contract. This contract defines five arrays, each containing 30 to 60 different short strings. In the contract functions, a random string from each array is selected and concatenated to form a complete name. In Cairo, we cannot define and initialize arrays directly in the contract. Currently, we are manually writing indexes using LegacyMap to achieve this.

![21693301815_ pic](https://github.com/starkware-libs/cairo/assets/28086155/fd3ffa32-041f-4160-b023-d23a2f6c7405)
![31693301850_ pic](https://github.com/starkware-libs/cairo/assets/28086155/548e8da3-8460-425c-8c48-8254ba9cf66c)
![41693301884_ pic](https://github.com/starkware-libs/cairo/assets/28086155/048816ff-0978-4df1-b5c7-43e37372e4f1)



## Proposition
We want to directly define string array data in the contract.

**Comment by orangeryang** [comment]
I want a map-like data structure (Array or Dict) that supports multiple levels of nesting, random write and can be used in memory, without relying on `LegacyMap` structures in storage.

**Reply by milancermak** [reply]
> map-like data structure (Array or Dict)

So an array or a dict?

Do you want to manipulate data only in memory or should it support contract storage?

**Comment by orangeryang** [comment]
Sorry I left some confusing things there. Actually I want a structure like 'Array<Array<T>>' that I can directly change the elements of whichever index I choose.
Cause Array just supports append() and pop() now due to the unchanged memory, so I mention the dict implements  the mutable feature there.
And it is good enough that it can support the memory I think.
So kind of you, have a good day!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by orangeryang** [comment]
I want '<<' and '>>' Operators. They provide left shift and right shift features with less gas cost.

**Comment by zhouop0** [comment]
In solidity contract, we can user block.number to get current block number, but i do not know how to get block number in cairo contract.

**Reply by tserg** [reply]
You can use `starknet::info::get_block_number()` (see [corelib](https://github.com/starkware-libs/cairo/blob/main/corelib/src/starknet/info.cairo)).

**Comment by zhouop0** [comment]
I know i can  use get_caller_address() to get  like msg.sender. but how can i get msg.value in cairo?

**Reply by zhouop0** [reply]
And in solidity, `type(unit).max`  if i am not wrong  that can get the max number for this type. Are there some function like that in cairo?

**Reply by swapnilraj** [reply]
There is no `msg.value` in Starknet. All value transfers are ERC20 transfers.

**Reply by zhouop0** [reply]
ok

**Comment by zhouop0** [comment]
how can i remove a element in Array by index? I can not find this method.

**Comment by zhouop0** [comment]
if we write legacyMap use the same key, whether  will replace the value??

**Comment by sunny-g** [comment]
Proof verification within a cairo program would be a massive step forward in building fully decentralized state machines!

**Comment by orangeryang** [comment]
I want Base64 utils that could support type like `Array<felt252>`

**Comment by cauchy48** [comment]
i wanna deploy contract on remix easily

**Comment by 0xNeshi** [comment]
**Make it possible to store functions inside variables.**

This will make it possible to, among other things, support more functional programming paradigms:
- **Pass functions as parameters to other functions** - e.g. will allow to define `map`, `filter`, `reduce` etc. on Arrays. Will allow iterator-related methods.
- **Return functions as results from other functions**
- **Define scoped (local) functions inside other functions**

**Comment by gaetbout** [comment]
Big DEV QOL improvement would be the possibility to make mutation testing. 
You'd run the tool, it would make some tweak in the code (for example changing a <= to a <) and expect a test to fail.
This could be a big improvement for checking some edge cases that are not correctly tested or detect a big if the tests are correct.
Existing tools are: 
- https://github.com/nfurfaro/hunter
- https://github.com/JoranHonig/vertigo

**Reply by gaetbout** [reply]
https://github.com/gaetbout/starknet-mutation-testing

**Comment by MullerEsposito** [comment]
Hey guys! Does the VSCode extension works on Windows? I'm struggling to make it work on Windows but it doesn't work. Someone use it on Windows?

**Reply by mkaput** [reply]
Hey! I don't think why it shouldn't work. But I can't help you if you don't tell me what's happening 😛 

Try running the Troubleshooting guide from extension's readme

**Reply by MullerEsposito** [reply]
Oh, ok! Let me try to explain. When I open the project in VSCode with WSL in remote mode, the extension works normally. In WSL I have the scarb installed through the asdf manager. So when I open the project through the WSL, the extension is able to communicate with the cairo-language-server.

![image](https://github.com/user-attachments/assets/c3a020ff-dd4f-4e7a-9055-13b8d3b0722f)

On other hand, when I open the project directly through the Windows, where I have the scarb installed manualy, cause asdf doesn't work in the Windows, the extension isn't able to communicate with the cairo-language-server.

![image](https://github.com/user-attachments/assets/7e6b2914-c63b-4827-9ac8-b02358681302)

If any moment you're want to make a video conference through Discord or any other platform I'm available.

**Reply by mkaput** [reply]
The extension is unable to find `scarb.exe` automatically on your Windows machine.

1. Do you have `scarb.exe` searchable from `PATH` environment variable? In other words: does running `scarb` work from PowerShell/Batch?
2. You can always manually put path to `scarb` in your VSCode configuration. Look up in preferences GUI.

When the extension gets to a valid Scarb installation, it should automatically find the `scarb-cairo-language-server.exe` binary which ships the LS.

**Comment by MullerEsposito** [comment]
1. Yep, my scarb is running in PowerShell.
![image](https://github.com/user-attachments/assets/da445c3f-7406-46eb-a4ed-b7f924c2293c)

 2. I changed manually the path to scarb in my VSCode configuration:
 
![image](https://github.com/user-attachments/assets/14aa5f4b-6cc7-49da-89fc-02feefaebb28)

This notification appeared on the screen:
![image](https://github.com/user-attachments/assets/a4ea81ad-7c07-4e4c-8f2c-e9358e3a8f13)

In the output this is the error message:
![image](https://github.com/user-attachments/assets/75988db8-3a58-4868-9fed-e98d0d9164cc)

**Reply by Draggu** [reply]
Hey! you have to specify path with binary `C:\Program Files\scarb\bin\scarb.exe` (same for LS).

**Reply by MullerEsposito** [reply]
@Draggu hey man! Thank you very much, it worked! I need to produce a tutorial for this.
